### PR TITLE
C front-end: initializer lists do not initialise anonymous members

### PR DESCRIPTION
**Pull request by tautschnig**
We already took care of this in another place in
c_typecheck_initializer, but did not handle initializer lists correctly.

The test is based on a case found by C-Reduce when starting from an
SV-COMP task.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by martin-cs — APPROVED** [review]
The small part that touches area I maintain looks OK.  I'm not sure the tests have enough context for me to really check what is going on with them.

File: regression/cbmc-library/memcpy-06/test.desc [hunk 1]
```diff
@@ -1,4 +1,4 @@
-CORE
+CORE pthread
```
Comment (kroening) [inline regression/cbmc-library/memcpy-06/test.desc]: Use `gcc-only` instead?

File: regression/ansi-c/struct5/test.desc [hunk 2]
```diff
@@ -1,6 +1,6 @@
 CORE
 main.c
-
+--i386-win32
 ^EXIT=0$
```
Comment (kroening) [inline regression/ansi-c/struct5/test.desc:4]: Seems unrelated?

Comment (tautschnig) [reply regression/ansi-c/struct5/test.desc:4]: No, this is a relevant change. We previously wrongly supported the code in this test on all platforms: it uses anonymous (but tagged) union members.

File: regression/cbmc-library/memcpy-06/test.desc (hunk 1 above)
Comment (tautschnig) [reply regression/cbmc-library/memcpy-06/test.desc]: Fixed.

**Comment by kroening** [comment]
Are zero-width bit-fields considered when processing initialiser lists?

**Comment by tautschnig** [comment]
@kroening:
> Are zero-width bit-fields considered when processing initialiser lists?

The C standard requires "The expression that specifies the width of a bit-field [...]. If the value is zero, the declaration shall have no declarator." GCC and Clang enforce this, we did not. I'll put together a PR to ensure our C front-end becomes equally strict.

File: src/ansi-c/c_typecheck_type.cpp [hunk 3]
```diff
@@ -948,6 +948,41 @@ void c_typecheck_baset::typecheck_compound_body(
...
+          else
+          {
+            // GCC and Clang ignore anything other than an untagged struct or
+            // union; we could print a warning, but there isn't any ambiguity in
+            // semantics here. Printing a warning could elevate this to an error
+            // when compiling code with goto-cc with -Werror.
+            if(
+              (new_component.type().id() != ID_struct_tag &&
+               new_component.type().id() != ID_union_tag) ||
+              follow(new_component.type()).find(ID_tag).is_not_nil())
```
Comment (kroening) [inline src/ansi-c/c_typecheck_type.cpp]: What does the follow do here?

Comment (tautschnig) [reply src/ansi-c/c_typecheck_type.cpp]: This is to distinguish untagged (anonymous) structs/unions from ones that are named.

Comment (tautschnig) [reply src/ansi-c/c_typecheck_type.cpp]: I have extended the comment above this code now.

Comment (kroening) [reply src/ansi-c/c_typecheck_type.cpp]: May I suggest using `to_tag_type` to make this clear.

Comment (tautschnig) [reply src/ansi-c/c_typecheck_type.cpp]: I've rewritten it to use `to_struct_tag_type` and `to_union_tag_type`.

**Review by peterschrammel — APPROVED** [review]
